`error[E0282]: type annotations needed` raised despite explicit type provided in function return signature

Here is a previous proved case. However, when I'm running this with latest Verus release, Verus raises the error `E0282: type annotations needed`, even though the return type is explicitly provided. 

Here is the Verus playground link:
https://play.verus-lang.org/?version=stable&mode=basic&edition=2021&gist=ec6952386e525c18dfe88e0d23315685

## PoC 
This issue occurs in the following code:
```rust
use vstd::prelude::*;

fn main() {
    // Write a function in Rust to find the sum of an array.

    assert_eq!(sum(&vec![1, 2, 3]), 6);
    assert_eq!(sum(&vec![15, 12, 13, 10]), 50);
    assert_eq!(sum(&vec![0, 1, 2]), 3);
}

verus! {

spec fn sum_to(arr: Seq<i64>) -> int
    decreases arr.len(),
{
    if arr.len() == 0 {
        0
    } else {
        sum_to(arr.drop_last()) + arr.last()
    }
}

fn sum(arr: &Vec<i64>) -> (sum: i128)
    ensures
        sum_to(arr@) == sum,
{
    let mut index = 0;
    let mut sum = 0i128;

    while index < arr.len()
        invariant
            0 <= index <= arr.len(),
            sum == sum_to(arr@.subrange(0, index as int)),
            forall|j: int|
                0 <= j <= index ==> (i64::MIN * index <= (sum_to(#[trigger] arr@.subrange(0, j)))
                    <= i64::MAX * index),
        decreases arr.len() - index,
    {
        assert(arr@.subrange(0, index as int) =~= arr@.subrange(0, (index + 1) as int).drop_last());
        sum = sum + arr[index] as i128;
        index += 1;
    }
    assert(arr@ == arr@.subrange(0, index as int));
    sum
}

} // verus!

```

## Error Message
```
error[E0282]: type annotations needed
  --> database/verusbench/code/benchmarks/MBPP/verified/task_id_798.rs:23:28
   |
23 | fn sum(arr: &Vec<i64>) -> (sum: i128)
   |                            ^^^
   |
help: consider giving this closure parameter an explicit type
   |
23 | fn sum(arr: &Vec<i64>) -> (sum: /* Type */: i128)
   |                               ++++++++++++

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0282`.
```

## Environment
```
$ database/vostd/code/tools/verus/source/target-verus/release/verus --version
Verus
  Version: 0.2025.08.23.652bf5c
  Profile: release
  Platform: linux_x86_64
  Toolchain: 1.88.0-x86_64-unknown-linux-gnu
```

--- a/source/builtin_macros/src/syntax.rs
+++ b/source/builtin_macros/src/syntax.rs
@@ -830,7 +830,28 @@ impl Visitor {
                 }
                 match std::mem::take(ret_opt) {
                     None => None,
-                    Some(ret) => Some((ret.1.clone(), ty.clone())),
+                    Some(ret) => {
+                        let original_pattern = ret.1.clone();
+                        let mut pattern = ret.1.clone();
+                        // Check if the pattern name conflicts with the function name
+                        let was_renamed = if let Pat::Ident(pat_ident) = &pattern {
+                            if pat_ident.ident.to_string() == sig.ident.to_string() {
+                                pattern = Pat::Verbatim(
+                                    quote_spanned! {pattern.span() => __verus_tmp_ret},
+                                );
+                                true
+                            } else {
+                                false
+                            }
+                        } else {
+                            false
+                        };
+                        Some((
+                            pattern,
+                            ty.clone(),
+                            if was_renamed { Some(original_pattern) } else { None },
+                        ))
+                    }
                 }
             }
         };
@@ -976,7 +997,7 @@ impl Visitor {
 
         let sig_span = sig.span().clone();
 
-        if let Some((p, _)) = &ret_pat {
+        if let Some((p, _, _)) = &ret_pat {
             if let Some(err_stmt) = check_verus_return_ident(p, &sig.inputs) {
                 stmts.push(err_stmt);
             }
@@ -984,8 +1005,8 @@ impl Visitor {
 
         let spec_stmts = self.take_sig_specs(
             &mut sig.spec,
-            ret_pat,
-            None,
+            ret_pat.as_ref().map(|(pat, ty, _)| (pat.clone(), ty.clone())),
+            ret_pat.as_ref().and_then(|(_, _, original_pat)| original_pat.clone()),
             sig_span,
             is_impl_fn,
             false,

--- a/source/rust_verify_test/tests/functions.rs
+++ b/source/rust_verify_test/tests/functions.rs
@@ -120,3 +120,14 @@ test_verify_one_file! {
         assert!(err.errors.iter().find(|p| p.message == "`impl Trait` in type aliases is unstable").is_some());
     }
 }
+
+test_verify_one_file! {
+    #[test] test_ensures_name_collision verus_code! {
+        fn test_fn() -> (test_fn: i32)
+            ensures
+                test_fn == 42,
+        {
+            42
+        }
+    } => Ok(())
+}

